rlkit/core/rl_algorithm.py

```diff
@@ -482,6 +495,128 @@ def get_extra_data_to_save(self, epoch):
+            path['goal'] = goal # goal
+
+        # save the paths for visualization, only useful for point mass
+        if self.dump_eval_paths:
+            logger.save_extra_data(paths, path='eval_trajectories/task{}-epoch{}-run{}'.format(idx, epoch, run))
+
+        return paths
+
+    def _do_eval(self, indices, epoch):
+        final_returns = []
+        online_returns = []
+        for idx in indices:
+            runs, all_rets = [], []
+            for r in range(self.num_evals):
+                paths = self.collect_paths(idx, epoch, r)
+                all_rets.append([eval_util.get_average_returns([p]) for p in paths])
+                runs.append(paths)
+            all_rets = np.mean(np.stack(all_rets), axis=0) # avg return per nth rollout
+            final_returns.append(all_rets[-1])
+            online_returns.append(all_rets)
+        return final_returns, online_returns
+
+    def evaluate(self, epoch):
+        statistics = OrderedDict()
+        statistics.update(self.eval_statistics)
+        self.eval_statistics = statistics
+
+        ### sample trajectories from prior for debugging / visualization
+        if self.dump_eval_paths:
+            prior_paths = []
+            for _ in range(100 // self.num_steps_per_eval):
+                # just want stochasticity of z, not the policy
+                prior_paths += self.eval_sampler.obtain_samples(deterministic=True, resample=np.inf)
+            logger.save_extra_data(prior_paths, path='eval_trajectories/prior-epoch{}'.format(epoch))
+
+        ### train tasks
+        # eval on a subset of train tasks for speed
+        indices = np.random.choice(self.train_tasks, len(self.eval_tasks))
+        dprint('evaluating on {} train tasks'.format(len(indices)))
+        ### eval train tasks with posterior sampled from the training replay buffer
+        train_returns = []
+        for idx in indices:
+            self.task_idx = idx
+            self.env.reset_task(idx)
+            paths = []
+            # TODO: put these lines into method evaluate_task(idx) that goes into sac.py?
+            for _ in range(10):
+                self.infer_posterior(idx)
+                paths += self.eval_sampler.obtain_samples(num_samples=self.max_path_length + 1, deterministic=True, resample=np.inf)
+            if self.sparse_rewards:
+                for p in paths:
+                    sparse_rewards = np.stack(e['sparse_reward'] for e in p['env_infos']).reshape(-1, 1)
+                    p['rewards'] = sparse_rewards
+            train_returns.append(eval_util.get_average_returns(paths))
+        train_returns = np.mean(train_returns)
+        ### eval train tasks with on-policy data to match eval of test tasks
+        train_final_returns, train_online_returns = self._do_eval(indices, epoch)
+        print('train online returns')
+        print(train_online_returns)
+
+        ### test tasks
+        # TOOD: should this be using the dprint in eval_util or pythons own dprint?
```

Comment:
I didn't know about python's `dprint` when I wrote this! We could use it, I just don't see how to turn it on/off?

Reply:
I think we are just using pythons dprint rjight now, since we would need to call like utils.dprint otherwise after importing utils or whatever file it was in.

Reply:
using the one in rl_utils